### PR TITLE
Calibration logic and uncertainties

### PR DESCRIPTION
**Pull request by liviajakob**
Changes to do with the GlaMBIE algorithm to combine timeseries as well ass added in uncertainties which were previously ignored.
- Added logic to average and calibrate a whole data catalogue of timeseries.
- Implemented uncertainties for each step in the GlaMBIE algorithm. Note that there are still some placeholders for steps where the Assessment Algorithm doesn't define the uncertainties yet which will need to be defined with the GlaMBIE consortium.

### CODE REVIEW

File: glambie/const/density_uncertainty.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+    Parameters
+    ----------
+    time_period_in_fractional_years : float
+        time period of desired density uncertainty
+
+    Returns
+    -------
+    int
+        Density uncertainty in kg/m^3
```
Comment (TristanGoss) [inline glambie/const/density_uncertainty.py:19]: can we be specific about what we mean by "uncertainty"? There are quite a few different ways of characterising uncertainty. Do we actually mean [Standard Error](https://en.wikipedia.org/wiki/Standard_error) here?

File: glambie/data/data_catalogue.py [hunk 2]
```diff
@@ -152,6 +168,96 @@ def datasets_are_same_unit(self):
...
+        -------
+        Tuple[Timeseries, DataCatalogue]
+            1. Timeseries object, containing the new combined solution
+            2. DataCatalogue, containing the catalogue the operation was performed on.
+               For remove_trend set to False, this is an exact copy of self.
+               For remove_trend set to True, this contains the altered datasets with the trends removed
+        """
+
+        if not self.datasets_are_same_unit():
+            warnings.warn("Warning: Trends within catalogue are not all the same unit. \
```
Comment (TristanGoss) [inline glambie/data/data_catalogue.py:201]: If you can't convert them all to the same units, shouldn't this be an error?

File: glambie/data/data_catalogue.py [hunk 3]
```diff
@@ -152,6 +168,96 @@ def datasets_are_same_unit(self):
...
+
+        df = reduce(lambda left, right: left.merge(right, how="outer", on=["start_dates", "end_dates"]), catalogue_dfs)
+        df = df.sort_values(by="start_dates")
+        start_dates, end_dates = np.array(df["start_dates"]), np.array(df["end_dates"])
+        mean_changes = np.array(df[df.columns.intersection(
+            df.filter(regex=("changes*")).columns.to_list())].mean(axis=1))
+
+        # UNCERTAINTIES -- more information is in GlaMBIE Assessment Algorithm document
+        # propagate observational errors
+        sigma_observational_errors = ((df.filter(regex=("errors*"))**2).sum(axis=1))**0.5
```
Comment (TristanGoss) [inline glambie/data/data_catalogue.py:231]: I can see from this math that yes, you are working with standard errors.

Ideally we'd do some renames to ensure that we never say "error" or "uncertainty" in isolation in code, as these are ambiguous terms.

I'm resonably sure that the approach here will result in an over-estimate, though, which is fine. I'd only worry if I saw something that would be likely to lead to an underestimate of the error.

File: glambie/data/data_catalogue_helpers.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    if not catalogue_with_trends.datasets_are_same_unit():
+        warnings.warn("Warning: Trends within catalogue are not all the same unit. This might lead to false results.")
+
+    # calibrate annual trends with longterm trend
+    calibrated_series = []
+    for ds in catalogue_with_trends.datasets:
+        # 1) calibrate timeseries
+        calibrated_s, dist_mat = calibrate_timeseries_with_trends(ds.data.as_dataframe(),
+                                                                  calibration_timeseries.data.as_dataframe())
+        # 2) caliculate mean calibration timeseries from all the different curves
```
Comment (TristanGoss) [inline glambie/data/data_catalogue_helpers.py:37]: typo

File: glambie/data/data_catalogue_helpers.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+                                           .as_dataframe().start_dates,
+                                           "end_dates": calibration_timeseries.data
+                                           .as_dataframe().end_dates, "changes": mean_calibrated_ts})
+        df_mean_calibrated_na_removed = df_mean_calibrated[~df_mean_calibrated["changes"].isna()]
+
+        # CALCULATE UNCERTAINTIES
+        # The uncertainty of a calibrated time series is calculated by combining the uncertainties of the anomalies
+        # and of the long-term trend
+        df_trends = ds.data.as_dataframe()
+        trend_errors = df_trends.errors  # remove na lines
```
Comment (TristanGoss) [inline glambie/data/data_catalogue_helpers.py:49]: the comment here should be on a new line, I think.

File: glambie/data/timeseries.py [hunk 6]
```diff
@@ -280,10 +284,21 @@ def convert_timeseries_to_unit_mwe(self, density_of_water: float = constants.DEN
...
                                                                                       density_of_ice=density_of_ice))
+                # Uncertainties
+                # First, convert elevation change error in m to mwe
+                object_copy.data.errors = np.array(meters_to_meters_water_equivalent(object_copy.data.errors,
+                                                                                     density_of_water=density_of_water,
+                                                                                     density_of_ice=density_of_ice))
+                # Second, include density uncertainty in error
+                density_unc = get_density_uncertainty_over_survey_period(self.data.max_temporal_resolution)
+                df = object_copy.data.as_dataframe()
+                # also see formula in Glambie Assessment Algorithm document, section 5.2 Homogenization of data
```
Comment (TristanGoss) [inline glambie/data/timeseries.py:299]: This is good, this is what I was looking for last time. Provided it's easy to trace back and forth, it's usually easy to spot implementation issues.

File: glambie/data/timeseries.py [hunk 7]
```diff
@@ -334,23 +349,41 @@ def convert_timeseries_to_unit_gt(self, include_area_change: bool = True,
...
         if self.unit == "gt":  # no conversion needed as already in gt
             return copy.deepcopy(self)
-        elif self.unit == "mwe":  # @TODO: convert uncertainties as well
+        elif self.unit == "mwe":
             if not include_area_change:
                 object_copy.data.changes = np.array(meters_water_equivalent_to_gigatonnes(
                     self.data.changes, area_km2=glacier_area, density_of_water=density_of_water))
-                return object_copy
+                # variables for uncertainty calculation
+                area_unc = glacier_area * self.region.area_uncertainty_percentage  # use individual glacier unc
```
Comment (TristanGoss) [inline glambie/data/timeseries.py:357]: Can you explain or reference what an "uncertainty percentage" is a bit more please? If it's something close to a hand-wave, we need to say so.

File: tests/data/test_data_catalogue.py [hunk 8]
```diff
@@ -114,3 +116,42 @@ def test_datasets_are_same_unit(example_catalogue):
...
+def test_average_timeseries_in_catalogue(example_catalogue_small):
+    example_catalogue_small.load_all_data()
+    example_catalogue_small.datasets.append(copy.deepcopy(example_catalogue_small.datasets[0]))
+
+    # since both timeseries are the same average should give back same as inputtimeseries
+    result_timeseries, _ = example_catalogue_small.average_timeseries_in_catalogue(remove_trend=False,
+                                                                                   add_trend_after_averaging=False)
+    assert np.array_equal(result_timeseries.data.changes, example_catalogue_small.datasets[0].data.changes)
+
+    # double the values for next test
```
Comment (TristanGoss) [inline tests/data/test_data_catalogue.py:138]: please split these up into individual unit tests. It should be obvious instantly when running them which are failing and which are passing, and the failing of one should not prevent the passing of another. This is only possible if they are written as separate tests.

File: glambie/data/data_catalogue.py (hunk 2 above)
Comment (TristanGoss) [reply glambie/data/data_catalogue.py:201]: I notice there's another one or two instances of this as well. If the production of a meaningful result is not guaranteed, we should throw an error.

**Review by TristanGoss — CHANGES_REQUESTED** [review]
A valuable addition and both a sensible size for and a sensible subject of a separate branch.

The constant use of deepcopy still feels wrong, but I know that the data volumes are very small, so it's not an issue.

I've left a few comments, but nothing major. The algorithm seems fine.

File: glambie/data/data_catalogue.py (hunk 3 above)
Comment (liviajakob) [reply glambie/data/data_catalogue.py:231]: Renaming to "uncertainties". In terms of the algorithm, there is no need to check as this is all defined by the GlaMBIE team plus advisory board... Except for a sanity check if the implementation makes sense :)

File: glambie/data/data_catalogue_helpers.py (hunk 4 above)
Comment (liviajakob) [reply glambie/data/data_catalogue_helpers.py:37]: done

File: glambie/data/data_catalogue_helpers.py (hunk 5 above)
Comment (liviajakob) [reply glambie/data/data_catalogue_helpers.py:49]: done

File: glambie/data/data_catalogue.py (hunk 2 above)
Comment (liviajakob) [reply glambie/data/data_catalogue.py:201]: raising AssertionErrors now

File: glambie/const/density_uncertainty.py (hunk 1 above)
Comment (liviajakob) [reply glambie/const/density_uncertainty.py:19]: as discussed in g-chat, will rename to uncertainties. ALso now refering to ATBD in the docstrings

File: tests/data/test_data_catalogue.py (hunk 8 above)
Comment (liviajakob) [reply tests/data/test_data_catalogue.py:138]: Done! This was just lazyness ;)

**Comment by liviajakob** [comment]
Thanks very much for the review!!
Requested changes were fairly simple & straight forward.
Hope you're happy with it :)

**Review by TristanGoss — APPROVED** [review]
Thanks for addressing the comments, happy to merge now.

**Pull request merged**